### DISCUSSION TITLE
Can I do customized value checking on the server side?

### DISCUSSION THREAD

**Discussion by HanShijia999** [keygen-sh/community#76] [Q&A] [answered]
We want to store the checksum of our source code on the Keygen server side.
At license activation time, the client-side app will calculate the checksum and send it to the server. Then the server can compare the checksum with the stored value and reject any request with a wrong checksum.

Is this possible?

Any help is appreciated!

**Comment by ezekg** [comment]
Not out of the box. I mean, you could use machine activation for this (e.g. make the machine fingerprint the checksum), but that's not really what it's meant for. Are you using Keygen to distribute your application? E.g. using releases [^0] and artifacts [^1].

[^0]: https://keygen.sh/docs/api/releases/
[^1]: https://keygen.sh/docs/api/artifacts/

**Reply by HanShijia999** [reply]
Hello Zeke,
Thank you for your comment. I'm surprised that we somehow reached the same solution.
We are using the offline licensing and we currently append the checksum after the machine fingerprint.
The downside of this solution is, a user can edit the source code before doing the activation and a wrong checksum will be saved.

Our App is still in the development phase, we haven't thought about distribution. Surprised to know we can even do that with Keygen.

Zeke, do you think I can store a checksum as some empty Artifact or some other object, then retrieve it and do checking on the client side?

**Reply by ezekg** [reply, marked as answer]
I was asking if you used Keygen to distribute because we're working on a new feature that may align with what you're asking for here: https://github.com/keygen-sh/keygen-api/issues/620. Essentially, you would be able to send a `scope.checksum` parameter during validation, and if the license doesn't have permission to access an artifact with that specific checksum, the validation will fail.

**Reply by HanShijia999** [reply]
Thank you Zeke!

**Reply by ezekg** [reply]
This is now available [via the `scope.checksum`](https://keygen.sh/docs/api/licenses/#licenses-validate-meta-scope.checksum) license validation scope.

**Reply by HanShijia999** [reply]
Thank the help Zeke.
The new feature fits our needs well!
